ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -189,72 +262,304 @@ def waypoints_cb(self, lane_msg):
+                # create the main part of the jmt curve
+                t = self.gen_point_in_jmt_curve(curpt, jmt_ptr, t)
+                if self.check_point(curpt) is True:
+                    recalc = True
+            else:
+                # The car has reached the point where it is near to target
+                # velocity
+                rospy.loginfo("{} beyond S = {} at ptr_id = {}".format
+                              (curpt.get_s(), JMT_instance.final_displacement,
+                               mod_ptr))
+                self.maintain_speed(mod_ptr, 1)
+        return recalc
+
+    def set_waypoints_velocity(self):
+        offset = 0           # offset in front of car to account for latency
+        safety_factor = 2.0  # increase slowing down distance
+        recalc = False       # indication that JMT calcs exceed bounds
+
+        dist_to_tl = self.get_dist_to_tl()
+
+        if self.waypoints[self.final_waypoints_start_ptr].get_v() == 0.0:
+            # we are stopped
+            offset = 0
+            stopping_distance = 0.0
+            self.state = 'stopped'
+        else:
+            stopping_distance = get_accel_distance(
+                self.waypoints[self.final_waypoints_start_ptr].get_v(),
+                0.0, self.default_accel/safety_factor)
+
+        rospy.loginfo("dist_to_tl = {}, stopping_dist = {}, state = {}"
+                      .format(dist_to_tl, stopping_distance, self.state))
+
+        # handle case where car is stopped at lights and light is red
+        if self.state == 'stopped' and dist_to_tl < self.dyn_tl_buffer:
+            self.set_stopped(self.final_waypoints_start_ptr, self.
+                             final_waypoints_start_ptr + self.lookahead_wps)
+
+        elif dist_to_tl < stopping_distance + self.dyn_tl_buffer or\
+                dist_to_tl < 20:    # added last check in to try to prevent
+                                    # flipping between accel and decel
+                                    # when approaching the light
+
+            if dist_to_tl < self.dyn_tl_buffer:
+                # small buffer from stop line - stop the car if in this area
+                rospy.loginfo("Now within tl_buffer = {}".format(self.dyn_tl_buffer))
+                self.set_stopped(self.final_waypoints_start_ptr, self.
+                                 lookahead_wps)
             else:
-                for ptr in range(start_ptr, final_end_ptr):
-                    # self.next_traffic_light_wp):
-                    mod_ptr = ptr % len(self.waypoints)
-                    chord_dist += self.distance(ptr, ptr+1)
-                    v = (1-chord_dist/distance_to_rl)*velocity
-                    if v < 0.9:
-                        v = 0.0
-                    self.waypoints[mod_ptr].twist.twist.linear.x = v
-                # end for
+                # reduce car velocity to 0.0 stoping before lights
+                recalc = self.produce_slowdown(self.final_waypoints_start_ptr,
```

Comment:
May want to add a log note that it's starting a deceleration.  I also just realized that the state is still called "speedup" so maybe it should be renamed to "moving" or split to "speedup" and "slowdown" for reference.

Reply:
will do